File: python3/70_climbing_stairs.py
import functools

# Dynamic approach with space optimization
# O(n) time
# O(1) space
class Solution1:
    def climbStairs(self, n: int) -> int:
        a, b = 1, 1
        for i in range(n):
            a, b = b, a + b
        return a

# Recursive approach with memoization
# O(n) time
# O(n) space
class Solution2:
    @functools.lru_cache()
    def climbStairs(self, n: int) -> int:
        return self.climbStairs(n - 1) + self.climbStairs(n - 2) if n > 2 else 2 if n == 2 else 1

File: javascript/70_climbing_stairs.js
/**
 * @param {number} n
 * @return {number}
 * O(n) time
 * O(1) space
 * Dynamic approach  
 */
var climbStairs = function(n) {
    for(var i = 3, F = [0,1,2]; i <= n; i++) F[i] = F[i - 1] + F[i - 2];
    return F[n] || 1;
};
